File: apps/api/app/scene_spatial/contracts.py
from __future__ import annotations

import hashlib
import json
import re
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


GEOMETRY_VERSION = "p2-backend-v0.2"
FIRST_PARTY_AUTHORITY = "FIRST_PARTY_BACKEND_GEOMETRY_SOLVER"


class SpatialPrecheckBody(BaseModel):
    model_config = ConfigDict(extra="allow")
    schema_version: str = "0.1.0"
    source_scan_id: str
    status: str
    authority: str
    reason: str = "UNSPECIFIED"
    routing: dict[str, Any] = Field(default_factory=dict)


class CameraModelEvidenceBody(BaseModel):
    model_config = ConfigDict(extra="allow")
    status: str
    focal_source: str = "UNKNOWN"
    principal_point_assumption: str = "UNKNOWN"
    distortion_assumption: str = "UNKNOWN"
    platform_device_profile: str = "UNKNOWN"
    confidence: float = 0.0


class SelectedGeometryFrameBody(BaseModel):
    model_config = ConfigDict(extra="allow")
    frame_id: str
    timestamp_ms: float
    relative_yaw_deg: float
    orientation_source: str
    width: int
    height: int
    source_width: int
    source_height: int
    working_width: int
    working_height: int
    encoded_bytes: int
    frame_sha256: str
    quality: float
    file_field: str


class SceneGeometryRequestBody(BaseModel):
    model_config = ConfigDict(extra="allow")
    schema_version: str = "0.1.0"
    geometry_request_id: str
    scan_id: str
    frame_set_hash: str
    geometry_version: str
    platform: str
    camera_model_evidence: CameraModelEvidenceBody
    client_precheck: SpatialPrecheckBody
    selected_geometry_frames: list[SelectedGeometryFrameBody]
    privacy: dict[str, Any]


def frame_sha256(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def frame_set_sha256(metadata: list[dict[str, Any]]) -> str:
    canonical = [[frame["frame_id"], str(frame["frame_sha256"]).lower()] for frame in metadata]
    payload = json.dumps(canonical, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    return hashlib.sha256(payload).hexdigest()


def validate_geometry_request(metadata: dict[str, Any], frames: list[bytes]) -> dict[str, Any]:
    parsed = SceneGeometryRequestBody.model_validate(metadata)
    request = parsed.model_dump()
    if parsed.geometry_version != GEOMETRY_VERSION:
        raise ValueError("GEOMETRY_VERSION_UNSUPPORTED")
    if not 1 <= len(parsed.geometry_request_id) <= 160 or not 1 <= len(parsed.scan_id) <= 160:
        raise ValueError("GEOMETRY_IDENTITY_INVALID")
    if re.fullmatch(r"[0-9a-fA-F]{64}", parsed.frame_set_hash) is None:
        raise ValueError("FRAME_SET_HASH_INVALID")
    if not 3 <= len(frames) <= 8 or len(frames) != len(parsed.selected_geometry_frames):
        raise ValueError("GEOMETRY_FRAME_COUNT_OUT_OF_BOUNDS")
    selected = request["selected_geometry_frames"]
    for frame, encoded in zip(selected, frames, strict=True):
        if frame_sha256(encoded) != frame["frame_sha256"].lower() or len(encoded) != frame["encoded_bytes"]:
            raise ValueError("FRAME_BINARY_HASH_MISMATCH")
        # Clients target a 640 px working edge; the accepted backend hard limit is 960 px.
        if max(frame["working_width"], frame["working_height"]) > 960 or min(frame["working_width"], frame["working_height"]) <= 0:
            raise ValueError("GEOMETRY_WORKING_RESOLUTION_OUT_OF_BOUNDS")
    if frame_set_sha256(selected) != parsed.frame_set_hash.lower():
        raise ValueError("FRAME_SET_HASH_MISMATCH")
    if parsed.camera_model_evidence.status not in {"KNOWN", "ESTIMATED_VALIDATED", "UNKNOWN"}:
        raise ValueError("CAMERA_MODEL_EVIDENCE_INVALID")
    if parsed.client_precheck.status not in {"UNRELIABLE", "NO_SIGNAL", "POSSIBLE"} or parsed.client_precheck.authority != "ROUTING_HINT_ONLY":
        raise ValueError("SPATIAL_PRECHECK_AUTHORITY_INVALID")
    expected_privacy = {
        "raw_video_upload": 0,
        "frame_stream_upload": 0,
        "provider_upload": 0,
        "luna_upload": 0,
        "selected_geometry_frame_upload": "FIRST_PARTY_BACKEND_ONLY",
    }
    if any(parsed.privacy.get(key) != value for key, value in expected_privacy.items()):
        raise ValueError("GEOMETRY_PRIVACY_CONTRACT_VIOLATION")
    return request

File: apps/client/src/sceneSpatial/contracts.ts
export type SceneSpatialMode='REAL'|'FAKE'|'REPLAY'
export type SpatialStatus='INSUFFICIENT'|'PARTIAL'|'USABLE'
export type SpatialPrecheckStatus='UNRELIABLE'|'NO_SIGNAL'|'POSSIBLE'
export type GeometryJobEvent='GEOMETRY_REQUESTED'|'SPATIAL_EVIDENCE_AVAILABLE'|'SPATIAL_EVIDENCE_INSUFFICIENT'|'GEOMETRY_FAILED'|'GEOMETRY_SUPERSEDED'

export interface SceneFrameObservation{
 frameId:string;relativeYawDeg:number;width:number;height:number;technicalUsability:number
 clutterScore:number;sharpnessScore:number;encodedBytes?:ArrayBuffer;filePath?:string;timestampMs:number
}
export interface SceneScanEvidenceV01{
 schema_version:'0.1.0';scan_id:string;captured_at:string
 coverage:{start_yaw_deg:number;end_yaw_deg:number;span_deg:number}
 frames:{frame_id:string;relative_yaw_deg:number;width:number;height:number;technical_usability:number;evidence_class:'FACT'}[]
 privacy:{raw_video_uploaded:0;frame_stream_uploaded:0;provider_calls:0;luna_calls:0}
}
export interface SceneFrameSetV01{schema_version:'0.1.0';source_scan_id:string;frame_refs:string[];raw_media_persisted:false;raw_media_uploaded:false}
export interface SceneDirectionMapV01{schema_version:'0.1.0';source_scan_id:string;basis:'RELATIVE_YAW';nodes:{frame_id:string;relative_yaw_deg:number;evidence_class:'FACT'}[];depth:'UNKNOWN';metric_geometry:'NOT_SUPPORTED'}
export interface CompositionAnchorCandidateV01{anchor_id:string;view_id:string;image_anchor:'LEFT_THIRD'|'CENTER'|'RIGHT_THIRD';confidence:number;authority:'IMAGE_PLANE_COMPOSITION_ANCHOR_ONLY'}
export interface ViewEvidenceV01{
 schema_version:'0.1.0';source_scan_id:string
 view_candidates:{view_id:string;frame_id:string;relative_yaw_deg:number;technical_usability:number;evidence_class:'CANDIDATE'}[]
 composition_anchors:CompositionAnchorCandidateV01[];authority:'P1_VIEW_CANDIDATE_EVIDENCE'
}
export interface SpatialPrecheckV01{schema_version:'0.1.0';source_scan_id:string;status:SpatialPrecheckStatus;authority:'ROUTING_HINT_ONLY';reason:string;routing:{backend_solve_recommended:boolean}}
export interface CameraModelEvidenceV01{status:'KNOWN'|'ESTIMATED_VALIDATED'|'UNKNOWN';focal_source:string;principal_point_assumption:string;distortion_assumption:string;platform_device_profile:string;confidence:number}
export interface SelectedGeometryFrameV01{frame_id:string;timestamp_ms:number;relative_yaw_deg:number;orientation_source:'DEVICE_ORIENTATION'|'CONTROLLED_FIXTURE';width:number;height:number;source_width:number;source_height:number;working_width:number;working_height:number;encoded_bytes:number;frame_sha256:string;quality:number;file_field:string}
export interface SceneGeometryRequestV01{
 schema_version:'0.1.0';geometry_request_id:string;scan_id:string;frame_set_hash:string;geometry_version:'p2-backend-v0.2';platform:'h5'|'wechat'|'fixture';camera_model_evidence:CameraModelEvidenceV01;client_precheck:SpatialPrecheckV01;selected_geometry_frames:SelectedGeometryFrameV01[]
 privacy:{raw_video_upload:0;frame_stream_upload:0;provider_upload:0;luna_upload:0;selected_geometry_frame_upload:'FIRST_PARTY_BACKEND_ONLY'}
}
export interface SpatialEvidenceV02{
 schema:'xfx.spatial-evidence';schema_version:'0.2.0';source_scan_id:string;status:SpatialStatus;status_authority:'FIRST_PARTY_BACKEND_GEOMETRY_SOLVER';confidence:number;geometry_type:'SPARSE_RELATIVE'|'UNKNOWN';metric_scale_available:false;limitations:string[];evidence_refs:string[];reason_codes:string[];diagnostics?:Record<string,unknown>
}
export interface SceneSpatialBundle{sceneScan:SceneScanEvidenceV01;frameSet:SceneFrameSetV01;directionMap:SceneDirectionMapV01;viewEvidence:ViewEvidenceV01;precheck:SpatialPrecheckV01}
export interface GeometryResult{geometryRequestId:string;spatialEvidence:SpatialEvidenceV02;cacheStatus:'CACHE_HIT'|'CACHE_MISS';providerMode:SceneSpatialMode}
export interface SceneSpatialCompletion{event:GeometryJobEvent;spatialEvidence:SpatialEvidenceV02|null;viewPathUsable:true;errorCode?:string}
